spec/app/ics-030-middleware/BEST_PRACTICES.md

```diff
@@ -0,0 +1,57 @@
+# Best Practices When Designing Middleware
+
+## Namespacing the Stack
+
+Each layer of the middleware should reserve a unique name in the stack so that it can key into specific information it needs in the version, packet data, or acknowledgement. Since this will not be in the portID which is shared across the stack, it will not be enforced by IBC. The Cosmos-SDK, for example, can accomplish uniqueness by using the ModuleName to reference the different middleware in the stack.
+
+## Version Negotiation For Middleware
+
+Middleware will only need to negotiate versions if they expect the counterparty chain to have compatible middleware on the stack as well. Middleware that can execute unilaterally **do not** need version negotiation.
+
+Middleware must take care to ensure that the application logic can execute its own version negotiation without interference from the nesting middleware. In order to do this, the middleware will format the version in a JSON-encoded string. Any information intended for the middleware will be keyed on the middleware namespace. The version intended for the underlying app will be keyed on `"app_version"`. The application version may as well be a JSON-encoded string, possibly including further middleware and app versions, if the application stack consists of multiple milddlewares wrapping a base application. The version values themselves may be more complex than static strings, they can be JSON encoded structs with multiple fields that the counterparties must agree on.  The format of the version string is as follows:
+
+```json
+{
+    "<middleware_name>": "<middleware_version_value>",
+    "app_version": "<application_version_value>",
+}
+```
+
+The `<middleware_name>` key in the JSON struct should be replaced by the actual name of the key for the corresponding middleware (e.g. `fee` for ICS-29 fee middleware).
+
+In the application callbacks, the middleware can unmarshal the version string and retrieve the middleware and application versions. It must do its own version negotiation on `<middleware_version_value>` and then hand over `<application_version_value>` to the nested application's callback. This is only relevant if the middleware expects a compatible counterparty middleware at the same level on the counterparty stack. Middleware that only executes on a single side of the channel MUST NOT modify the channel version.
+
+## Packet Data Structuring For Middleware
+
+Packet senders may choose to send input data not just to the base application but also the middlewares in the stack. The packet data should be structured in the same way that the middleware stack is; i.e. nested from the top level middleware to the base application.
+
+Similar to the version negotiation, the `app_packet_data` may be marshalling packet data for underlying middleware as well.
+
+```json
+{
+    "<middleware_name>": "<middleware_packet_data>",
+    "app_packet_data": "<application_packet_data>",
+}
+```
```

Comment:
The `OnRecvPacket` callbacks for ibc apps accept the arg `channeltypes.Packet`. That means we would have to reassign `packet.Data` bytes to `AppPacketData` before calling the next handler in the stack. 

If we consider the `OnRecvPacket` callback of [ics29 fee middleware](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/ibc_middleware.go#L217) and imagine that there is some middleware specific packet data encoded within the packet we would have to do something like:

```go
var packetData types.FeePacketData
if err := cdc.Unmarshal(packet.GetData(), packetData); err != nil {
    // handle err or pass through to underlying app
}

// do whatever logic with fee specific packet data
handleFeePacketData(packetData.FeePacketData)

// reassign packet data bytes before passing packet to next handler
packet.Data = packetData.AppPacketData

return im.app.OnRecvPacket(ctx, packet, relayer)
```

Would this be the intended usage here?

Reply:
Hmm yes, this is a good point. Correct this will be necessary, but again only for middleware that wish to modify the packet data

Reply:
The `OnRecvPacket` packet interface would need to be updated within the spec if we are going to allow packet data nesting as a best practice?